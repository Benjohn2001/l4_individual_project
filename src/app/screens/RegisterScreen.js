import React, { useState, useRef } from "react";
import {
  Text,
  View,
  Image,
  TextInput,
  SafeAreaView,
  Alert,
} from "react-native";
import { createUserWithEmailAndPassword } from "firebase/auth";
import {
  getDatabase,
  ref,
  set,
  orderByChild,
  query,
  equalTo,
  get,
} from "firebase/database";
import PressableIcon from "../components/PressableIcon";
import COLOURS from "../assets/colours";
import AppButtonPurple from "../components/AppButtonPurple";
import { auth } from "../../firebase";

function RegisterScreen({ navigation }) {
  const [firstName, setFirstName] = useState("");
  const [lastName, setLastName] = useState("");
  const [userName, setUserName] = useState("");
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");

  const db = getDatabase();

  const registerUser = async () => {
    const unamesRef = query(
      ref(db, "/users"),
      orderByChild("userName"),
      equalTo(userName)
    );

    if (
      firstName === "" ||
      lastName === "" ||
      userName === "" ||
      email === "" ||
      password === ""
    ) {
      Alert.alert("Incomplete Form", "One or more fields are empty");
    } else if (password.length < 8 || password.search(/[A-Z]/) < 0) {
      Alert.alert(
        "Improper Password",
        "Password must be at least 8 characters and contain at least 1 upper-case character"
      );
    } else if (userName !== "") {
      const snapshot = await get(unamesRef);
      if (snapshot.val() !== null) {
        Alert.alert("Username taken", "Please select a new username");
      } else {
        createUserWithEmailAndPassword(auth, email, password)
          .then((userCred) => {
            set(ref(db, `users/${userCred.user.uid}`), {
              firstName,
              lastName,
              userName,
              email,
              bio: "Add a bio!"
            });
            setFirstName("");
            setLastName("");
            setUserName("");
            setEmail("");
            setPassword("");
            navigation.navigate("SignInScreen");
          })
          .catch((error) => {
            if (error.code === "auth/invalid-email") {
              Alert.alert(
                "Invalid E-mail",
                "Please enter a valid email address"
              );
            } else if (error.code === "auth/email-already-in-use") {
              Alert.alert("Email taken", "This email is already in use");
            }
          });
      }
    }
  };

  const lastNameRef = useRef();
  const usernameRef = useRef();
  const emailRef = useRef();
  const passwordRef = useRef();

  return (
    <SafeAreaView className="flex-1 bg-primaryPurple">
      <View className="pt-10 pl-5">
        <PressableIcon
          onPress={() => {
            navigation.navigate("SignInScreen");
          }}
          icon="arrow-left"
          size={40}
          color="black"
        />
      </View>
      <View className="flex-row justify-center items-center">
        <Image className="w-44 h-44" source={require("../assets/clock.png")} />
        <View className="items-center">
          <Text className="font-bold text-2xl">Weasley Clock</Text>
          <Text className="text-base">Join today. It's free!</Text>
        </View>
      </View>
      <View className="items-center">
        <TextInput
          className="bg-secondaryPurple my-4 w-80 h-12 rounded-md"
          placeholder="First Name"
          underlineColorAndroid="transparent"
          cursorColor={COLOURS.darkerPurple}
          value={firstName}
          onChangeText={(val) => setFirstName(val)}
          returnKeyType="next"
          onSubmitEditing={() => {
            lastNameRef.current.focus();
          }}
          blurOnSubmit={false}
        />
        <TextInput
          className="bg-secondaryPurple my-4 w-80 h-12 rounded-md"
          placeholder="Last Name"
          underlineColorAndroid="transparent"
          cursorColor={COLOURS.darkerPurple}
          value={lastName}
          onChangeText={(val) => setLastName(val)}
          ref={lastNameRef}
          returnKeyType="next"
          onSubmitEditing={() => {
            usernameRef.current.focus();
          }}
          blurOnSubmit={false}
        />
        <TextInput
          className="bg-secondaryPurple my-4 w-80 h-12 rounded-md"
          placeholder="Username"
          underlineColorAndroid="transparent"
          cursorColor={COLOURS.darkerPurple}
          value={userName}
          onChangeText={(val) => setUserName(val)}
          ref={usernameRef}
          returnKeyType="next"
          onSubmitEditing={() => {
            emailRef.current.focus();
          }}
          autoCapitalize="none"
          blurOnSubmit={false}
        />
        <TextInput
          className="bg-secondaryPurple my-4 w-80 h-12 rounded-md"
          placeholder="Email"
          underlineColorAndroid="transparent"
          cursorColor={COLOURS.darkerPurple}
          value={email}
          onChangeText={(val) => setEmail(val)}
          ref={emailRef}
          returnKeyType="next"
          onSubmitEditing={() => {
            passwordRef.current.focus();
          }}
          blurOnSubmit={false}
          autoCapitalize="none"
        />
        <TextInput
          className="bg-secondaryPurple mt-4 w-80 h-12 rounded-md"
          placeholder="Password"
          underlineColorAndroid="transparent"
          cursorColor={COLOURS.darkerPurple}
          secureTextEntry
          value={password}
          onChangeText={(val) => setPassword(val)}
          ref={passwordRef}
          autoCapitalize="none"
        />
        <Text className="text-sm text-left pt-2 pb-2 text-gray-500">
          Password must be at least 8 characters and {"\n"}
          contain at least 1 upper-case character
        </Text>
        <AppButtonPurple
          title="Submit"
          onPress={() => {
            registerUser();
          }}
        />
      </View>
    </SafeAreaView>
  );
}
export default RegisterScreen;
