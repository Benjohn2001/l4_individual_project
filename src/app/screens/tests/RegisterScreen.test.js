import renderer from "react-test-renderer";
import { render, screen, fireEvent } from "@testing-library/react-native";
import { Alert } from "react-native";
import RegisterScreen from "../RegisterScreen";

jest.mock("firebase/database", () => ({
  getDatabase: jest.fn().mockReturnThis(),
  push: jest.fn().mockReturnThis(),
  ref: jest.fn().mockReturnThis(),
  set: jest.fn().mockReturnThis(),
  query: jest.fn(()=>{Promise.resolve()})
}));

jest.mock("firebase/storage", () => ({
  getStorage: jest.fn().mockReturnThis(),
}));

const navigation = {
  navigate: jest.fn(),
  push: jest.fn(),
};

jest.mock("firebase/app", () => ({
  auth: jest.fn().mockReturnThis(),
  initializeApp: jest.fn().mockReturnThis(),
}));

jest.mock("firebase/auth", () => ({
  auth: jest.fn().mockReturnThis(),
  createUserWithEmailAndPassword: () => ({
    then: () => ({
      catch: jest.fn(),
    }),
  }),
  getAuth: () => ({
    signOut: () => ({
      then: () => ({
        catch: jest.fn(),
      }),
    }),
  }),
}));

test("Check RegisterScreen  working correctly", () => {
  const { getByTestId } = render(<RegisterScreen navigation={navigation} />);

  fireEvent.changeText(getByTestId("fnameT"), "abc");
  fireEvent.changeText(getByTestId("snameT"), "abc");
  fireEvent.changeText(getByTestId("unameT"), "abc");
  fireEvent.changeText(getByTestId("emailT"), "abc");
  fireEvent.changeText(getByTestId("passT"), "12345678B");
  fireEvent.press(screen.getByText("Submit"));

  const domTree = renderer.create(<RegisterScreen />).toJSON();
  expect(domTree).toMatchSnapshot();
});


test("Check invalidpass  working correctly", () => {
    const { getByTestId } = render(<RegisterScreen navigation={navigation} />);
    jest.spyOn(Alert, "alert");
  
    fireEvent.changeText(getByTestId("fnameT"), "abc");
    fireEvent(getByTestId("fnameT"), 'submitEditing')
    fireEvent.changeText(getByTestId("snameT"), "abc");
    fireEvent(getByTestId("snameT"), 'submitEditing')
    fireEvent.changeText(getByTestId("unameT"), "abc");
    fireEvent(getByTestId("unameT"), 'submitEditing')
    fireEvent.changeText(getByTestId("emailT"), "abc");
    fireEvent(getByTestId("emailT"), 'submitEditing')
    fireEvent.changeText(getByTestId("passT"), "1238B");
    fireEvent.press(screen.getByText("Submit"));
    // expect(Alert.alert).toHaveBeenCalledTimes(1);
    
  });

test("Check RegisterScreen no login data working correctly", () => {
  jest.spyOn(Alert, "alert");
  render(<RegisterScreen navigation={navigation} />);

  fireEvent.press(screen.getByText("Submit"));
//   expect(Alert.alert).toHaveBeenCalledTimes(1);
});

test("Check register nav to signin", () => {
  const { getByTestId } = render(<RegisterScreen navigation={navigation} />);

  fireEvent.press(getByTestId("pressableIcon"));
  expect(navigation.navigate).toBeCalledWith("SignInScreen");
});
