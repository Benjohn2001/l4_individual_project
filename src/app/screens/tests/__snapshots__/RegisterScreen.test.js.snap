// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check RegisterScreen  working correctly 1`] = `
<RCTScrollView
  style={
    Array [
      Object {
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
      },
      Object {
        "backgroundColor": "#C6C4FF",
      },
    ]
  }
>
  <View>
    <RCTSafeAreaView
      emulateUnlessSupported={true}
    >
      <View
        style={
          Array [
            Object {
              "paddingTop": 40,
            },
            Object {
              "paddingLeft": 20,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
              },
            ]
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            testID="pressableIcon"
          >
            <Text />
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            Object {
              "justifyContent": "center",
            },
            Object {
              "alignItems": "center",
            },
          ]
        }
      >
        <Image
          source={
            Object {
              "uri": "uri",
            }
          }
          style={
            Array [
              Object {
                "width": 176,
              },
              Object {
                "height": 176,
              },
            ]
          }
        />
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontWeight": "700",
                },
                Object {
                  "fontSize": 24,
                  "lineHeight": 32,
                },
              ]
            }
          >
            Weasley Clock
          </Text>
          <Text
            style={
              Array [
                Object {
                  "fontSize": 16,
                  "lineHeight": 24,
                },
              ]
            }
          >
            Join today. It's free!
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
            },
          ]
        }
      >
        <TextInput
          blurOnSubmit={false}
          cursorColor="#6B4EFF"
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          placeholder="First Name"
          returnKeyType="next"
          style={
            Array [
              Object {
                "backgroundColor": "#E7E7FF",
              },
              Object {
                "marginBottom": 16,
                "marginTop": 16,
              },
              Object {
                "width": 320,
              },
              Object {
                "height": 48,
              },
              Object {
                "borderBottomLeftRadius": 6,
                "borderBottomRightRadius": 6,
                "borderTopLeftRadius": 6,
                "borderTopRightRadius": 6,
              },
            ]
          }
          testID="fnameT"
          underlineColorAndroid="transparent"
          value=""
        />
        <TextInput
          blurOnSubmit={false}
          cursorColor="#6B4EFF"
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          placeholder="Last Name"
          returnKeyType="next"
          style={
            Array [
              Object {
                "backgroundColor": "#E7E7FF",
              },
              Object {
                "marginBottom": 16,
                "marginTop": 16,
              },
              Object {
                "width": 320,
              },
              Object {
                "height": 48,
              },
              Object {
                "borderBottomLeftRadius": 6,
                "borderBottomRightRadius": 6,
                "borderTopLeftRadius": 6,
                "borderTopRightRadius": 6,
              },
            ]
          }
          testID="snameT"
          underlineColorAndroid="transparent"
          value=""
        />
        <TextInput
          autoCapitalize="none"
          blurOnSubmit={false}
          cursorColor="#6B4EFF"
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          placeholder="Username"
          returnKeyType="next"
          style={
            Array [
              Object {
                "backgroundColor": "#E7E7FF",
              },
              Object {
                "marginBottom": 16,
                "marginTop": 16,
              },
              Object {
                "width": 320,
              },
              Object {
                "height": 48,
              },
              Object {
                "borderBottomLeftRadius": 6,
                "borderBottomRightRadius": 6,
                "borderTopLeftRadius": 6,
                "borderTopRightRadius": 6,
              },
            ]
          }
          testID="unameT"
          underlineColorAndroid="transparent"
          value=""
        />
        <TextInput
          autoCapitalize="none"
          blurOnSubmit={false}
          cursorColor="#6B4EFF"
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          placeholder="Email"
          returnKeyType="next"
          style={
            Array [
              Object {
                "backgroundColor": "#E7E7FF",
              },
              Object {
                "marginBottom": 16,
                "marginTop": 16,
              },
              Object {
                "width": 320,
              },
              Object {
                "height": 48,
              },
              Object {
                "borderBottomLeftRadius": 6,
                "borderBottomRightRadius": 6,
                "borderTopLeftRadius": 6,
                "borderTopRightRadius": 6,
              },
            ]
          }
          testID="emailT"
          underlineColorAndroid="transparent"
          value=""
        />
        <TextInput
          autoCapitalize="none"
          cursorColor="#6B4EFF"
          onChangeText={[Function]}
          placeholder="Password"
          secureTextEntry={true}
          style={
            Array [
              Object {
                "backgroundColor": "#E7E7FF",
              },
              Object {
                "marginTop": 16,
              },
              Object {
                "width": 320,
              },
              Object {
                "height": 48,
              },
              Object {
                "borderBottomLeftRadius": 6,
                "borderBottomRightRadius": 6,
                "borderTopLeftRadius": 6,
                "borderTopRightRadius": 6,
              },
            ]
          }
          testID="passT"
          underlineColorAndroid="transparent"
          value=""
        />
        <Text
          style={
            Array [
              Object {
                "fontSize": 14,
                "lineHeight": 20,
              },
              Object {
                "textAlign": "left",
              },
              Object {
                "paddingTop": 8,
              },
              Object {
                "paddingBottom": 8,
              },
              Object {
                "color": "#6b7280",
              },
            ]
          }
        >
          Password must be at least 8 characters and 
          

          contain at least 1 upper-case character
        </Text>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#6B4EFF",
              "borderBottomLeftRadius": 9999,
              "borderBottomRightRadius": 9999,
              "borderTopLeftRadius": 9999,
              "borderTopRightRadius": 9999,
              "height": 48,
              "marginLeft": 6,
              "marginRight": 6,
              "opacity": 1,
              "paddingTop": 12,
              "width": "83.333333%",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#fff",
                },
                Object {
                  "textAlign": "center",
                },
              ]
            }
          >
            Submit
          </Text>
        </View>
      </View>
    </RCTSafeAreaView>
  </View>
</RCTScrollView>
`;
